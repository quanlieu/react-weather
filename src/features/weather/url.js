export const url = {
  LOCATION_SEARCH_QUERY: 'http://localhost:4000/location/search/?query=',
  LOCATION: 'http://localhost:4000/location/',
}