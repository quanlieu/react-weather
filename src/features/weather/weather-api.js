const locationSearchApi = {
  searchLocationsByName: async ({ name }) => {
    try {
      const response = await fetch(`http://localhost:4000/location/search/?query=${name}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      })
      const locations = await response.json();
      return locations
    } catch (error) {
      
    }
  },
}

export default locationSearchApi
