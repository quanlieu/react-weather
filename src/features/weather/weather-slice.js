import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import _get from 'lodash/get'
import weathersApi from './weather-api'

export const initialState = {
  locations: [],
  locationsFetchStatus: 'idle',
  locationsFetchId: 0,
  weatherData: [],
  weatherDataFetchStatus: 'idle',
};

export const searchLocationsByName = createAsyncThunk(
  'weather/searchLocationsByName',
  async (name, thunkAPI) => {
    try {
      const state = thunkAPI.getState()
      const currentFetchId = selectLocationsFetchId(state)
      const response = await weathersApi.searchLocationsByName({ name });
      if (currentFetchId !== selectLocationsFetchId(thunkAPI.getState())) {
        // In case a request is called before another request but resolve after
        return;
      }

      return response.map(location => ({
        value: location.woeid,
        label: location.title,
      }));
    } catch (error) {
      thunkAPI.rejectWithValue('Error');
      throw error;
    }
  }
);

export const fetchWeatherData = createAsyncThunk(
  'weather/fetchWeatherData',
  async (woeId, thunkAPI) => {
    try {
      const response = await weathersApi.fetchWeatherDataByWoeId({ woeId });
      return _get(response, 'consolidated_weather', []).map(data => {
        const id = _get(data, 'id');
        const date = _get(data, 'applicable_date');
        const minTemp = _get(data, 'min_temp');
        const maxTemp = _get(data, 'max_temp');
        return {
          id, date, minTemp, maxTemp,
        }
      });
    } catch (error) {
      thunkAPI.rejectWithValue('Error');
      throw error;
    }
  }
);

export const weatherSlice = createSlice({
  name: 'weather',
  initialState,
  reducers: {
  },
  extraReducers: (builder) => {
    builder
      .addCase(searchLocationsByName.pending, (state) => {
        state.locationsFetchId += 1;
        state.locationsFetchStatus = 'loading';
        state.locations = [];
      })
      .addCase(searchLocationsByName.fulfilled, (state, action) => {
        state.locationsFetchStatus = 'idle';
        state.locations = _get(action, 'payload', state.locations);
      })
      .addCase(fetchWeatherData.pending, (state) => {
        state.weatherDataFetchStatus = 'loading';
        state.weatherData = [];
      })
      .addCase(fetchWeatherData.fulfilled, (state, action) => {
        state.weatherDataFetchStatus = 'idle';
        state.weatherData = _get(action, 'payload');
      });
  },
});

// Selectors
export const selectLocations = (state) => state.weather.locations;
export const selectLocationsFetchId = (state) => state.weather.locationsFetchId;
export const selectWeatherData = (state) => state.weather.weatherData;
export const selectIsFetchingWeatherData = (state) => {
  const { weatherData, weatherDataFetchStatus } = state.weather;
  return weatherData.length === 0 && weatherDataFetchStatus === 'loading';
};

export default weatherSlice.reducer;
