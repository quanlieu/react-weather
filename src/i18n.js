import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';

const resources = {
  en: {
    translation: {
      weatherSearch: 'Search',
      weatherLocationSelect: 'Please select a location',
      today: 'Today',
      min: 'Min',
      max: 'Max',
    },
  },
  vi: {
    translation: {
      search: 'Tìm kiếm',
      weatherLocationSelect: 'Vui lòng chọn nơi cần xem',
      today: 'Hôm nay',
      min: 'Thấp nhất',
      max: 'Cao nhất',
    },
  },
};

i18n
  .use(initReactI18next)
  .init({
    resources,
    lng: 'en',
    keySeparator: false,
    interpolation: {
      escapeValue: false,
    },
  });

export default i18n;
