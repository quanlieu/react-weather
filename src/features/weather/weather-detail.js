import { Card, Typography } from 'antd';
import i18n from 'i18next';
import { useTranslation } from 'react-i18next';

import dateHelpers from '../../helpers/date-helpers'

export const WeatherDetail = ({ date, minTemp, maxTemp }) => {
  const { t } = useTranslation();
  let dateString = new Intl.DateTimeFormat(i18n.language, { weekday: 'long' })
    .format(Date.parse(date));
  if (dateHelpers.isToday(date)) {
    dateString = t('today');
  }
  return (
    <Card
      title={(
        <Typography.Title level={5}>
          {dateString}
        </Typography.Title>
      )}
    >
      <p>{t('min')}: {Math.round(minTemp * 10) / 10}&deg;C</p>
      <p>{t('max')}: {Math.round(maxTemp * 10) / 10}&deg;C</p>
    </Card>
  );
};
