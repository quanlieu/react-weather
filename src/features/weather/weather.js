import { Fragment, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import {
  Row, Col, Divider, Skeleton, Typography, Switch,
} from 'antd';
import i18n from 'i18next';
import { useTranslation } from 'react-i18next';

import {
  searchLocationsByName,
  fetchWeatherData,
  selectLocations,
  selectWeatherData,
  selectIsFetchingWeatherData,
} from './weather-slice';
import { DebounceSelect } from '../../components/debounce-select/debounce-select'
import { WeatherDetail } from './weather-detail'

export const Weather = props => {
  const { t } = useTranslation();
  const locations = useSelector(selectLocations);
  const weatherData = useSelector(selectWeatherData);
  const isFetchingWeatherData = useSelector(selectIsFetchingWeatherData);
  const dispatch = useDispatch();
  const handleSearch = useCallback(
    value => dispatch(searchLocationsByName(value)),
    [dispatch],
  );
  const handleChangeLocation = useCallback(
    location => dispatch(fetchWeatherData(location.value)),
    [dispatch],
  );
  const handleChangeLanguage = useCallback(
    checked => i18n.changeLanguage(checked ? 'en' : 'vi'),
    [],
  );

  return (
    <Fragment>
      <Row>
        <Col xs={24} sm={12} md={8} lg={6} xl={5}>
          <DebounceSelect
            placeholder={t('search')}
            onSearch={handleSearch}
            onChange={handleChangeLocation}
            options={locations}
            dropdownMatchSelectWidth={false}
            style={{
              width: '80%',
            }}
          />
          <div
            style={{
              width: '20%',
              textAlign: 'right',
              display: 'inline-block',
            }}
          >
            <Switch
              checkedChildren="En"
              unCheckedChildren="Vi"
              defaultChecked
              onChange={handleChangeLanguage}
            />
          </div>
        </Col>
      </Row>
      <Divider orientation="left" />
      <Skeleton loading={isFetchingWeatherData}>
        <Row
          gutter={[
            { xs: 8, sm: 8, md: 12, lg: 12, xl: 16},
            { xs: 8, sm: 8, md: 12, lg: 12, xl: 16},
          ]}
        >
          {!weatherData.length && (
            <Col>
              <Typography.Title level={3}>
                {t('weatherLocationSelect')}
              </Typography.Title>
            </Col>
          )}
          {weatherData.map(data => (
            <Col xs={12} md={8} xl={4} key={data.id}>
              <WeatherDetail {...data} />
            </Col>)
          )}
        </Row>
      </Skeleton>
    </Fragment>
  );
}
