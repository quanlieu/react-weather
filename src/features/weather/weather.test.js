import * as reactRedux from 'react-redux';
import i18n from 'i18next';
import { render, screen, fireEvent } from '../../test-utils'
import { Weather } from './weather';
import * as weatherSlice from './weather-slice';

jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: v => v,
  }),
}))

jest.mock('antd', () => ({
  Row: ({ children }) => <div>{children}</div>,
  Col: ({ children }) => <div>{children}</div>,
  Divider: () => 'Divider',
  Skeleton: ({ children }) => <div>{children}</div>,
  Typography: {
    Title: ({ children }) => <div>{children}</div>,
  },
  Switch: ({ onChange }) => <input data-testid='switch-language' onChange={onChange} />,
}))

jest.mock('../../components/debounce-select/debounce-select', () => ({
  DebounceSelect: (props) => (
    <div>
      DebounceSelect
      <input data-testid='on-search' onChange={props.onSearch} />
      <input data-testid='on-change' onChange={props.onChange} />
    </div>
  ),
}))

jest.mock('./weather-detail', () => ({
  WeatherDetail: () => <div>WeatherDetail</div>,
}))

const initialState = {
  weather: weatherSlice.initialState,
}

const initialStateWithWeatherDetail = {
  weather: {
    ...initialState.weather,
    weatherData: [{
      id: '1', date: '2021-05-01', minTemp: 30, maxTemp: 35,
    }, {
      id: '2', date: '2021-05-02', minTemp: 30, maxTemp: 35,
    }, {
      id: '3', date: '2021-05-03', minTemp: 30, maxTemp: 35,
    }, {
      id: '4', date: '2021-05-04', minTemp: 30, maxTemp: 35,
    }, {
      id: '5', date: '2021-05-05', minTemp: 30, maxTemp: 35,
    }, {
      id: '6', date: '2021-05-06', minTemp: 30, maxTemp: 35,
    }],
  },
}
describe('Weather component', () => {
  test('Should render default', () => {
    const wrapper = render(<Weather />, { initialState })
    expect(wrapper).toMatchSnapshot();
    expect(screen.getByText('DebounceSelect')).toBeInTheDocument();
    expect(screen.queryByText('WeatherDetail')).toBe(null);
  })

  test('Should render WeatherDetail', () => {
    const wrapper = render(<Weather />, { initialState: initialStateWithWeatherDetail });
    expect(wrapper).toMatchSnapshot();
    expect(screen.getAllByText('WeatherDetail').length).toBe(6);
  })

  test('Should call searchLocationsByName', () => {
    const dispatch = jest.fn()
    reactRedux.useDispatch = jest.fn(() => dispatch)
    weatherSlice.searchLocationsByName = jest.fn(() => ({
      type: 'ACTION',
    }));
    render(<Weather />, { initialState: initialStateWithWeatherDetail });
    fireEvent.change(screen.getByTestId('on-search'), { target: { value: 'wil' } });
    expect(weatherSlice.searchLocationsByName).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith({
      type: 'ACTION',
    });
  })

  test('Should call fetchWeatherData', () => {
    const dispatch = jest.fn()
    reactRedux.useDispatch = jest.fn(() => dispatch)
    weatherSlice.fetchWeatherData = jest.fn(() => ({
      type: 'ACTION',
    }));
    render(<Weather />, { initialState: initialStateWithWeatherDetail });
    fireEvent.change(screen.getByTestId('on-change'), { target: { value: 'wil' } });
    expect(weatherSlice.fetchWeatherData).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledWith({
      type: 'ACTION',
    });
  })

  test('Should change language', () => {
    i18n.changeLanguage = jest.fn()
    render(<Weather />, { initialState: initialStateWithWeatherDetail });
    fireEvent.change(screen.getByTestId('switch-language'), { target: { value: true } });
    expect(i18n.changeLanguage).toHaveBeenCalled();
  })
});
