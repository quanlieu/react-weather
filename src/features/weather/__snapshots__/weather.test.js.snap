// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather component Should render WeatherDetail 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div>
          <div>
            DebounceSelect
            <input
              data-testid="on-search"
            />
            <input
              data-testid="on-change"
            />
          </div>
          <div
            style="width: 20%; text-align: right; display: inline-block;"
          >
            <input
              data-testid="switch-language"
            />
          </div>
        </div>
      </div>
      Divider
      <div>
        <div>
          <div>
            <div>
              WeatherDetail
            </div>
          </div>
          <div>
            <div>
              WeatherDetail
            </div>
          </div>
          <div>
            <div>
              WeatherDetail
            </div>
          </div>
          <div>
            <div>
              WeatherDetail
            </div>
          </div>
          <div>
            <div>
              WeatherDetail
            </div>
          </div>
          <div>
            <div>
              WeatherDetail
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div>
        <div>
          DebounceSelect
          <input
            data-testid="on-search"
          />
          <input
            data-testid="on-change"
          />
        </div>
        <div
          style="width: 20%; text-align: right; display: inline-block;"
        >
          <input
            data-testid="switch-language"
          />
        </div>
      </div>
    </div>
    Divider
    <div>
      <div>
        <div>
          <div>
            WeatherDetail
          </div>
        </div>
        <div>
          <div>
            WeatherDetail
          </div>
        </div>
        <div>
          <div>
            WeatherDetail
          </div>
        </div>
        <div>
          <div>
            WeatherDetail
          </div>
        </div>
        <div>
          <div>
            WeatherDetail
          </div>
        </div>
        <div>
          <div>
            WeatherDetail
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Weather component Should render default 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div>
          <div>
            DebounceSelect
            <input
              data-testid="on-search"
            />
            <input
              data-testid="on-change"
            />
          </div>
          <div
            style="width: 20%; text-align: right; display: inline-block;"
          >
            <input
              data-testid="switch-language"
            />
          </div>
        </div>
      </div>
      Divider
      <div>
        <div>
          <div>
            <div>
              weatherLocationSelect
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div>
        <div>
          DebounceSelect
          <input
            data-testid="on-search"
          />
          <input
            data-testid="on-change"
          />
        </div>
        <div
          style="width: 20%; text-align: right; display: inline-block;"
        >
          <input
            data-testid="switch-language"
          />
        </div>
      </div>
    </div>
    Divider
    <div>
      <div>
        <div>
          <div>
            weatherLocationSelect
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
